genie-web/src/main/java/com/netflix/genie/web/apis/rest/v3/controllers/JobRestController.java

```diff
@@ -692,54 +693,69 @@ public void getJobOutput(
-                    this.restTemplate.execute(
-                        forwardHost + JOB_API_BASE_PATH + id + "/output/" + path,
-                        HttpMethod.GET,
-                        forwardRequest -> copyRequestHeaders(request, forwardRequest),
-                        (ResponseExtractor<Void>) forwardResponse -> {
-                            response.setStatus(forwardResponse.getStatusCode().value());
-                            copyResponseHeaders(response, forwardResponse);
-                            // Documentation I could find pointed to the HttpEntity reading the bytes off
-                            // the stream so this should resolve memory problems if the file returned is large
-                            ByteStreams.copy(forwardResponse.getBody(), response.getOutputStream());
-                            return null;
-                        }
-                    );
-                } catch (final HttpClientErrorException.NotFound e) {
-                    throw new GenieNotFoundException("Not Found (via: " + forwardHost + ")", e);
-                } catch (final HttpStatusCodeException e) {
-                    throw new GenieException(
-                        e.getStatusCode().value(),
-                        "Proxied request failed: " + e.getMessage(),
-                        e
-                    );
-                } catch (final Exception e) {
-                    log.error("Failed getting the remote job output from {}. Error: {}", forwardHost, e.getMessage());
-                    throw new GenieServerException("Proxied request error:" + e.getMessage(), e);
-                }
+        if (archiveStatus == ArchiveStatus.PENDING) {
+            final boolean isV4 = this.persistenceService.isV4(id);
+            final String jobHostname;
+            try {
+                jobHostname = this.getJobOwnerHostname(id, isV4);
+            } catch (NotFoundException e) {
+                throw new GenieServerException("Failed to route request", e);
+            }
+
+            final boolean shouldForward = !this.hostname.equals(jobHostname);
+            final boolean canForward = forwardedFrom == null && this.jobsProperties.getForwarding().isEnabled();
 
-                //No need to search on this node
+            if (shouldForward && canForward) {
+                // Forward request to another node
+                forwardRequest(id, path, jobHostname, request, response);
                 return;
+            } else if (!canForward && shouldForward) {
+                // Should forward but can't
+                throw new GenieServerException("Job files are not local, but forwarding is disabled");
             }
         }
 
+        // In any other case, delegate the request to the service
         log.debug("Fetching requested resource \"{}\" for job \"{}\"", path, id);
         this.jobDirectoryServerService.serveResource(id, baseUrl, path, request, response);
     }
 
+    private void forwardRequest(
```

Comment:
This is generically named `forwardRequest` but we also forward kill requests and it looks like this only happens for output requests